### ISSUE TITLE
Use of IGrainStorage for storing state in Orleans.EventSourcing.EventStoreStorage.LogViewAdaptor

### ISSUE THREAD

**Issue by willg1983** [hongliyu2002/Orleans.EventStore#3]
It seems odd to snapshot the event sourced grain state to IGrainStorage on every write, what was the rational behind that? 

It would be nice to have that behaviour be opt-in, and ideally something the grain can explicitly ask for - to match natural business requirements for snapshots (e.g. end of period) or for technical optimisation (e.g. on grain deactivation to speed up activation later).

**Comment by hongliyu2002** [comment]
You're right. It looks like we need to do a bit of improvement on this package. The initial idea was to keep the state up-to-date, which is the simplest and least error-prone approach. However, it seems that we need to learn some common practices for stateful grains, and provide methods such as WriteState, ReadState, and ClearState. This way, users can decide when to write/read the state to/from storage on their own. 
Maybe I can also add some methods like RebuildSnapshot...

**Comment by hongliyu2002** [comment]
You are correct. It appears that we need to make some improvements to this package. The original concept was to maintain an up-to-date state in the state storage, which is the most straightforward and least error-prone method.

However, it seems that JournaledGrain does not support state operations like stateful grain does, such as WriteState, ReadState, and ClearState.  So the user cannot decide when to write/read snapshot state to/from the storage.

Therefore, we need to add some configuration to LogViewAdaptor to enable features like AutoSave or Rebuild the snapshot, and make it compatible with JournaledGrain.

**Comment by willg1983** [comment]
One approach could be to define an interface which could receive an interface that allows that level of control. 

Something like 

```
// implemented by the ILogViewAdaptor
public interface ISnapshotProvider
{
 Task SnapshotStateAsync(bool truncateLog = false);
}

// to be implemented by the grain
public interface ISnapshotGrain 
{
   InstallSnapshotControl(ISnapshotProvider snapshotProvider)
}
```

OnPreActivate the log view adaptor would attempt to cast _host to `ISnapshotGrain` and if successful pass itself to `InstallSnapshotControl`

the grain could then store the reference to `
ISnapshotProvider` in a field and call `SnapshotStateAsync` which the logViewAdaptor would implement and use `IGrainStorage` to write a versioned snapshot of the grain state, and optionally truncate the log. 

Ideally the interfaces would be defined in Microsoft.Orleans.EventSourcing and perhaps even JournaledGrain or another derived type could implement the interface and provide a simple Snapshot() method for derived types to call.

**Comment by willg1983** [comment]
I've also raised this with the Orleans community here https://github.com/dotnet/orleans/issues/8408 which would move the API surface for this into JournaledGrain in Microsoft.Orleans.EventSourcing, let me know your thoughts.
